packages/libs/deer/src/lib.rs

```diff
@@ -6,4 +6,212 @@
 #![allow(clippy::module_name_repetitions)]
 #![forbid(unsafe_code)]
 
+use alloc::string::String;
+
+use error_stack::{IntoReport, Result};
+use num_traits::ToPrimitive;
+
+pub use crate::{error::Error, number::Number};
+
+mod error;
+mod number;
+
 extern crate alloc;
+
+/// Lookup of Type
+///
+/// This is used as the signal for the type that is currently present, [`Deserialize`]
+/// implementations may use this to special case certain values.
+#[derive(Debug, Copy, Clone)]
+pub enum PrimitiveType {
+    /// null type
+    Null,
+    /// bool type
+    Bool,
+    /// number type
+    Number,
+    /// string type
+    String,
+    /// array type
+    Array,
+    /// object type
+    Object,
+}
+
+macro_rules! derive_from_number {
+    ([$($name:ident; $method:ident),*]) => {
+        $(derive_from_number!(#internal, $name; $method);)*
+    };
+
+    (#internal, $name:ident; $method:ident) => {
+        /// Automatically implemented convenience method, which uses [`Self::number`] to extract
+        /// a value of the primitive type, will otherwise error out.
+        ///
+        /// # Errors
+        ///
+        /// Current value is either not a number or wasn't able to be casted to the primitive type
+        fn $name(self) -> Result<$name, Self::Error> {
+            self.number()?
+                .$method()
+                .ok_or_else(|| Self::Error::message("provided value too large or too small"))
+                .into_report()
+        }
+    };
+}
+
+/// A **data format** that can deserialize any data structure which is supported by deer.
+///
+/// This traits defines the deserialization half, while [`Deserialize`] is used to use the
+/// deserializer to convert that data into the instance of a type, or fail if that was not possible.
+///
+/// `deer`s focus is on simplicity, therefore the data model is a lot less specific than ones used
+/// by other crates (like [`serde`]).
+///
+/// The data model consists of the following types:
+///
+/// * 4 primitives:
+///     * `null` (equivalent to [`None`] or `()`)
+///     * `bool` (equivalent to [`true`], [`false`])
+///     * `number` (equivalent to `1u8`, `257i16`, ...)
+///     * `string` (equivalent to `"Hello World!"`)
```

Comment:
I don't think that "equivalent" is the correct term here.

Reply:
fixed in [c8a8e45](https://github.com/hashintel/hash/pull/1170/commits/c8a8e452b4d72c95b18b136f85172f35768d77b8)